contribs/github-bot/bot.go

```diff
@@ -0,0 +1,188 @@
+	p "github.com/gnolang/gno/contribs/github-bot/params"
+	"github.com/gnolang/gno/contribs/github-bot/utils"
+	"github.com/google/go-github/v64/github"
+	"github.com/xlab/treeprint"
+)
+
+func execBot(params *p.Params) error {
+	// Init GitHub API client
+	gh := client.New(params)
+	defer gh.Close()
+
+	// Handle comment update, if any
+	if err := handleCommentUpdate(gh); err != nil {
+		return fmt.Errorf("comment update handling failed: %w", err)
+	}
+
+	// Retrieve a slice of pull requests to process
+	var (
+		prs []*github.PullRequest
+		err error
+	)
+
+	// If requested, retrieve all open pull requests
+	if params.PrAll {
+		opts := &github.PullRequestListOptions{
+			State:     "open",
+			Sort:      "updated",
+			Direction: "desc",
+		}
+
+		prs, _, err = gh.Client.PullRequests.List(gh.Ctx, gh.Owner, gh.Repo, opts)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve all open pull requests: %w", err)
+		}
+
+		// Otherwise, retrieve only specified pull request(s) (flag or GitHub Action context)
+	} else {
+		prs = make([]*github.PullRequest, len(params.PrNums))
+		for i, prNum := range params.PrNums {
+			pr, _, err := gh.Client.PullRequests.Get(gh.Ctx, gh.Owner, gh.Repo, prNum)
+			if err != nil {
+				return fmt.Errorf("unable to retrieve specified pull request (%d): %w", prNum, err)
+			}
+			prs[i] = pr
+		}
+	}
+
+	if len(prs) > 1 {
+		prNums := make([]int, len(prs))
+		for i, pr := range prs {
+			prNums[i] = pr.GetNumber()
+		}
+
+		gh.Logger.Infof("%d pull requests to process: %v\n", len(prNums), prNums)
+	}
+
+	// Process all pull requests in parallel
+	autoRules, manualRules := config(gh)
+	var wg sync.WaitGroup
+	wg.Add(len(prs))
+
+	// Used in dry-run mode to log cleanly from different goroutines
+	logMutex := sync.Mutex{}
+
+	for _, pr := range prs {
+		go func(pr *github.PullRequest) {
+			defer wg.Done()
+			commentContent := CommentContent{}
+			commentContent.allSatisfied = true
+
+			// Iterate over all automatic rules in config
+			for _, autoRule := range autoRules {
+				ifDetails := treeprint.NewWithRoot(fmt.Sprintf("%s Condition met", utils.StatusSuccess))
+
+				// Check if conditions of this rule are met by this PR
+				if autoRule.If.IsMet(pr, ifDetails) {
+					c := AutoContent{Description: autoRule.Description, Satisfied: false}
+					thenDetails := treeprint.NewWithRoot(fmt.Sprintf("%s Requirement not satisfied", utils.StatusFail))
+
+					// Check if requirements of this rule are satisfied by this PR
+					if autoRule.Then.IsSatisfied(pr, thenDetails) {
+						thenDetails.SetValue(fmt.Sprintf("%s Requirement satisfied", utils.StatusSuccess))
+						c.Satisfied = true
+					} else {
+						commentContent.allSatisfied = false
+					}
+
+					c.ConditionDetails = ifDetails.String()
+					c.RequirementDetails = thenDetails.String()
+					commentContent.AutoRules = append(commentContent.AutoRules, c)
+				}
+			}
+
+			// Iterate over all manual rules in config
+			for _, manualRule := range manualRules {
```

Comment:
It's better not to do this since some `Then` / `Requirement` of the automatic checks can sometimes modify something (for example, assigning a user, applying a label, etc.). This can therefore affect the validity of an `If` / `Condition` of a manual check.